rpcs3/rpcs3qt/log_frame.cpp

```diff
@@ -316,24 +348,108 @@ void log_frame::UpdateUI()
 		return QString(buf.data());
 	};
 
+	auto write_to_log = [&](const QString& text, const QColor& color)
+	{
+		// save old log state
+		QScrollBar *sb = m_log->verticalScrollBar();
+		bool isMax = sb->value() == sb->maximum();
+		int sb_pos = sb->value();
+		int sel_pos = m_log->textCursor().position();
+		int sel_start = m_log->textCursor().selectionStart();
+		int sel_end = m_log->textCursor().selectionEnd();
+
+		// clear selection or else it will get colorized as well
+		QTextCursor c = m_log->textCursor();
+		c.clearSelection();
+		m_log->setTextCursor(c);
+
+		QString suffix;
+		bool isSame = text == m_old_text;
+
+		// create counter suffix and remove recurring line if needed
+		if (m_stack_log)
+		{
+			if (isSame)
+			{
+				m_log_counter++;
+				suffix = QString(" x%1").arg(m_log_counter);
+				m_log->moveCursor(QTextCursor::End, QTextCursor::MoveAnchor);
+				m_log->moveCursor(QTextCursor::StartOfBlock, QTextCursor::MoveAnchor);
+				m_log->moveCursor(QTextCursor::End, QTextCursor::KeepAnchor);
+				m_log->textCursor().removeSelectedText();
+				m_log->textCursor().deletePreviousChar();
+			}
+			else
+			{
+				m_log_counter = 1;
+				m_old_text = text;
+			}
+		}
+
+		// add actual log message
+		m_log->setTextColor(color);
+		m_log->append(text);
+
+		// add counter suffix if needed
+		if (m_stack_log && isSame)
+		{
+			m_log->setTextColor(m_color_stack);
+			m_log->insertPlainText(suffix);
+		}
+
+		// if we mark text from right to left we need to swap sides (start is always smaller than end)
+		if (sel_pos < sel_end)
+		{
+			std::swap(sel_start, sel_end);
+		}
+
+		// reset old text cursor and selection
+		c.setPosition(sel_start);
+		c.setPosition(sel_end, QTextCursor::KeepAnchor);
+		m_log->setTextCursor(c);
+
+		// set scrollbar to max means auto-scroll
+		sb->setValue(isMax ? sb->maximum() : sb_pos);
+	};
+
 	const auto start = steady_clock::now();
 
 	// Check TTY logs
 
+	int tty_count = -1;
+	static QColor tty_col = m_color[static_cast<int>(logs::level::notice)];
+
 	while (const u64 size = std::min<u64>(buf.size(), m_tty_file.size() - m_tty_file.pos()))
 	{
 		QString text = get_utf8(m_tty_file, size);
 
-		// Hackily used the state of the check..  be better if I actually stored this value.
-		if (m_TTYAct->isChecked())
+		if (m_log_tty)
 		{
-			text.chop(1); // remove newline since Qt automatically adds a newline.
+			if (m_show_tty_in_log && ++tty_count == 0)
+			{
+				m_log->setTextColor(tty_col);
+				m_log->append("--- TTY begin ---");
```

Comment:
Consider prefixing the tty log text similarly to log channels

Reply:
feel free to pr a better solution. I think this PR is dead anyway

Reply:
I'm pretty sure neko is just occupied and/or doesn't feel like merging prs atm

Reply:
Regarding this PR, I was opposing some TTY changes from the start, so it became a complicated problem.

Reply:
I see. Well, input was added at some point, so it should suffice yes?

Reply:
I could change it to just have the input lineedit in tty.
But that will be useless until someone hooks it up